<div style="direction:rtl;line-height:300%;"> <font face="XB Zar">
<div align=center>
    <div style="font-size:14px">
    به نام خدا
    </div><br>
    <strong><div style="font-size:50px; color: blue">
    ci - hw2
    </div></strong>
    <br>
    <div style="font-size:20px; color: #ff6600">
    مائده نادهی 9712762238
    </div>
</div>
<hr>

<div style="direction:rtl; font-size: 17px; font-family: 'XB Zar'; line-height: 200%">
<div style="font-size: 22px; color: #ff6600; font-weight: bold">
فهرست محتویات
</div>         
<ul style="margin-right: 0;">
    <li><a href="#section1">preprocessing</a></li>
    <li><a href="#section2">decision tree</a></li>
</ul>
</div>

In [1]:
import pandas as pd
import numpy as np
import time
from sklearn.tree import DecisionTreeClassifier
from sklearn import metrics

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
!unzip '/content/drive/MyDrive/MNIST dataset - train.zip'
!unzip '/content/drive/MyDrive/MNIST dataset - test.zip'

Archive:  /content/drive/MyDrive/MNIST dataset - train.zip
  inflating: train_label.csv         
  inflating: train_set.csv           
Archive:  /content/drive/MyDrive/MNIST dataset - test.zip
  inflating: test_label.csv          
  inflating: test_set.csv            


In [9]:
train_label = pd.read_csv("/content/train_label.csv")
train_label.rename(columns = {"5" : "actual_labels"}, inplace=True)
train_label = train_label["actual_labels"]
train_set = pd.read_csv("/content/train_set.csv")
train_set.set_axis([i for i in range(0,784)], axis=1, inplace=True)
test_set = pd.read_csv("/content/test_set.csv")
test_set.set_axis([i for i in range(0,784)], axis=1, inplace=True)
test_label = pd.read_csv("/content/test_label.csv")
test_label = test_label["7"]

<div id="section1" style="direction:rtl; font-size: 17px; font-family: 'XB Zar'; line-height: 200%">
<div style="font-size: 22px; color: #ff6600; font-weight: bold">
preprocessing
</div>
<hr>
<div>
ابتدا بررسی میکنیم داده ی غیر معتبر NULL و NAN نداشته باشیم :
</div>
</div>

In [10]:
# how many null?
print(test_set.isnull().sum().isnull().sum(axis = 0))
print(train_set.isnull().sum().isnull().sum(axis = 0))

0
0


<div style="direction:rtl; font-size: 17px; font-family: 'XB Zar'; line-height: 200%">
<div>
   الگوریتم درخت تصمیم را برای دو نوع دیتاست اجرا میکنیم: 1. دیتاست بدون تغییر 2. دیتاست با تعداد فیچر های کمتر به این صورت که فیچر هایی را که انحراف معیار آنها کمتر از 0.4 را حذف میکنیم (چون پیکسل های بسیاری در همه ی عکس ها مقدار تقریبا یکسانی داشتند به طور مثال حدودا 60 پیکسل اطراف همگی مقدار 0 داشتند)
</div>
</div>

In [67]:
con = train_set.std() > 0.4
xtrain2 = train_set.loc[:, con]
xtest2 = test_set.loc[:, con]
xtrain2.shape

(59999, 191)

<div style="direction:rtl; font-size: 17px; font-family: 'XB Zar'; line-height: 200%">
<div>
    همانطور که نشان داده شده است از 784 فیچر 191 فیچر باقی ماندند.(انحراف معیار بیشتر از 0.4 داشته اند)
</div>
</div>

In [11]:
xtrain = train_set
ytrain = train_label
xtest = test_set
ytest = test_label

In [12]:
def eval_metrics(description, y_true, y_pred):
  print(description)
  print('Accuracy: {:.2f}'.format(metrics.accuracy_score(y_true, y_pred)))
  print('Rand Index: {:.2f}'.format(metrics.rand_score(y_true, y_pred)))
  print('------------------------')

In [74]:
def decisionTree(xtrain, ytrain, xtest, ytest, criterion, max_depth=None, max_features=None):
  dtc = DecisionTreeClassifier(criterion=criterion, max_depth=max_depth, max_features=max_features, random_state=0)
  print(dtc)
  t1 = time.time()
  dtc.fit(xtrain, ytrain)
  t2 = time.time()
  print('time : ', t2 - t1)
  y_pred = dtc.predict(xtest)
  eval_metrics("criterion: " + str(criterion) + " max_depth: " + str(max_depth) + " max_features: " + str(max_features), ytest, y_pred)

<div id="section2" style="direction:rtl; font-size: 17px; font-family: 'XB Zar'; line-height: 200%">
<div style="font-size: 22px; color: #ff6600; font-weight: bold">
decision tree
</div>
</div>

In [64]:
decisionTree(xtrain2, ytrain, xtest2, ytest, criterion = 'entropy')
decisionTree(xtrain2, ytrain, xtest2, ytest, criterion = 'gini')

DecisionTreeClassifier(criterion='entropy', random_state=0)
time :  11.891611337661743
criterion: entropy max_depth: None max_features: None
Accuracy: 0.88
Rand Index: 0.96
------------------------
DecisionTreeClassifier(random_state=0)
time :  9.581332206726074
criterion: gini max_depth: None max_features: None
Accuracy: 0.86
Rand Index: 0.95
------------------------


In [16]:
decisionTree(xtrain, ytrain, xtest, ytest, criterion = 'entropy')
decisionTree(xtrain, ytrain, xtest, ytest, criterion = 'gini')

DecisionTreeClassifier(criterion='entropy', random_state=0)
time :  22.710269689559937
criterion: entropy max_depth: None max_features: None
Accuracy: 0.88
Rand Index: 0.96
------------------------
DecisionTreeClassifier(random_state=0)
time :  23.624271392822266
criterion: gini max_depth: None max_features: None
Accuracy: 0.88
Rand Index: 0.96
------------------------


<div style="direction:rtl; font-size: 17px; font-family: 'XB Zar'; line-height: 200%">
<div>
    همانطور که مشاهده میشود حذف فیچر های با انحراف معیار کمتر از 0.4 موجب نصف شدن زمان اجرا میشود بدون تغییر در Accuracy 
    <br>
    نوع criterion هم تفاوت چندانی ایجاد نمیکند بنابراین ازین جای کار به بعد entropy را برای criterion درنظر میگیریم که همان معیار انتخاب بهترین فیچر است که در کلاس تدریس شده   
</div>
</div>

In [68]:
decisionTree(xtrain2, ytrain, xtest2, ytest, criterion = 'entropy', max_depth=14)

DecisionTreeClassifier(criterion='entropy', max_depth=14, random_state=0)
time :  12.511010646820068
criterion: entropy max_depth: 14 max_features: None
Accuracy: 0.88
Rand Index: 0.96
------------------------


In [25]:
decisionTree(xtrain, ytrain, xtest, ytest, criterion = 'entropy', max_depth=14)
decisionTree(xtrain, ytrain, xtest, ytest, criterion = 'entropy')
decisionTree(xtrain, ytrain, xtest, ytest, criterion = 'entropy', max_depth=9)

DecisionTreeClassifier(criterion='entropy', max_depth=14, random_state=0)
time :  21.260932445526123
criterion: entropy max_depth: 14 max_features: None
Accuracy: 0.89
Rand Index: 0.96
------------------------
DecisionTreeClassifier(criterion='entropy', random_state=0)
time :  22.443628311157227
criterion: entropy max_depth: None max_features: None
Accuracy: 0.88
Rand Index: 0.96
------------------------
DecisionTreeClassifier(criterion='entropy', max_depth=9, random_state=0)
time :  15.076143503189087
criterion: entropy max_depth: 9 max_features: None
Accuracy: 0.86
Rand Index: 0.95
------------------------


<div style="direction:rtl; font-size: 17px; font-family: 'XB Zar'; line-height: 200%">
<div>
    با افزایش مقدار max_depth به 14 شاهد افزایش Accuracy تا <span style="color: red">89 درصد</span> 
    هستیم
</div>
</div>

In [23]:
decisionTree(xtrain, ytrain, xtest, ytest, criterion = 'entropy', max_features='log2')
decisionTree(xtrain, ytrain, xtest, ytest, criterion = 'entropy', max_features='sqrt')
decisionTree(xtrain, ytrain, xtest, ytest, criterion = 'entropy')

DecisionTreeClassifier(criterion='entropy', max_features='log2', random_state=0)
time :  0.6413168907165527
criterion: entropy max_depth: None max_features: log2
Accuracy: 0.80
Rand Index: 0.93
------------------------
DecisionTreeClassifier(criterion='entropy', max_features='sqrt', random_state=0)
time :  1.285703182220459
criterion: entropy max_depth: None max_features: sqrt
Accuracy: 0.84
Rand Index: 0.94
------------------------
DecisionTreeClassifier(criterion='entropy', random_state=0)
time :  22.51461958885193
criterion: entropy max_depth: None max_features: None
Accuracy: 0.88
Rand Index: 0.96
------------------------


In [20]:
!pip install -U scikit-learn

     |████████████████████████████████| 23.2 MB 126 kB/s 
  Attempting uninstall: scikit-learn
    Found existing installation: scikit-learn 0.22.2.post1
    Uninstalling scikit-learn-0.22.2.post1:
      Successfully uninstalled scikit-learn-0.22.2.post1
